File: others/best_time_to_buy_and_sell_stock_2.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let ret=0,cur=0,tmp=0
    for(let i=prices.length-1;i>=0;i--){
        if(prices[i]>cur){
            cur = prices[i]
        }
        else{
            ret += cur-prices[i]
            cur = prices[i]
        }
    }
    return ret
};
console.log(maxProfit([7,1,5,3,6,4]))
console.log(maxProfit([1,2,3,4,5]))
console.log(maxProfit([7,6,4,3,1]))

File: others/best_time_to_buy_and_sell_stock_2.go
package main

import (
    "fmt"
    "math"
)

func maxProfit(prices []int) int {
    ret := 0
    for i:=0;i<len(prices)-1;i++ {
        ret += int(math.Max(0,float64(prices[i+1]-prices[i])))
    }
    return ret
}

func main(){
    t1 := []int{7,1,5,3,6,4}
    t2 := []int{1,2,3,4,5}
    t3 := []int{7,6,4,3,1}
    fmt.Println(maxProfit(t1))
    fmt.Println(maxProfit(t2))
    fmt.Println(maxProfit(t3))
}
